tests/testthat/test-type-date-time.R

```diff
@@ -153,9 +207,17 @@ test_that("safe casts work as expected", {
   expect_error(vec_cast(list(date), date), class = "vctrs_error_incompatible_type")
 })
 
+test_that("date - datetime cast can be roundtripped", {
+  date <- as.Date("2018-01-01")
+  datetime <- as.POSIXct("2018-01-01", tz = "America/New_York")
+
+  expect_identical(vec_cast(vec_cast(date, datetime), date), date)
+  expect_identical(vec_cast(vec_cast(datetime, date), datetime), datetime)
+})
+
 test_that("lossy casts generate error", {
   date <- as.Date("2018-01-01")
-  datetime <- as.POSIXct(date) + c(0, 3600)
+  datetime <- as.POSIXct(as.character(date)) + c(0, 3600)
```

Comment:
Why do you need this?

Reply:
It goes along with one of the bug fixes. Compare:

```r
date <- as.Date("2018-01-01")

as.POSIXct(date)
#> [1] "2017-12-31 19:00:00 EST"

as.POSIXct(as.character(date))
#> [1] "2018-01-01 EST"
```

This test expects that when you allow a lossy cast of the `datetime` to `date`, that it gives identical results to `vec_c(date, date)`. If we don't cast to character first to ignore time zone shifts, then with the newly corrected behavior casting to date will give `"2017-12-31"` which won't match `date`

Reply:
Oh of course 🙄